### DISCUSSION TITLE
Run a template script from within a template

### DISCUSSION THREAD

**Discussion by daamiian** [SilentVoid13/Templater#1119] [Help] [answered]
I made a script with the help of chatGPT to run the templater select template modal but modified to display the folders of the Utilities/Templates folder first, and allow the user to select the template once navigated to the base folder, and then the file is created, the user is asked for a title and the file is moved to it's default folder. Basically a script to automate creating a naming a file, maybe there was an easier way.

The reason why I made a script for selecting a template is because I was able to add creating a naming the new file more seamlessly, whereas before I had to create a file, name it, open the templater modal then select the template. 

I have another script that I modified from a helpful user on my last discussion post, and it moves the file depending on some conditions (if it belongs to a project or course folder, it will be placed there, otherwise its default folder is kept. It allows me to have standalone lecture notes inside my Input Notes folder while having course-specific lecture notes kept in the course folder. 

Here is the desired workflow:

1. Create a note using the `select template` script I made, it will pre-prompt for a template, create the note using the selected template, prompt for a title, and focus on the note. *NOTE: Inside the template I selected, there is a move function to move the file to its default folder (e.g. a Lecture note would be moved to Input Notes), this is to pre-sort the note before asking if it belongs to a collection.*
3. [Not implemented] I want to now check if the note is a part of a project or course, so I want to run the `move file to collection` script to check if the note is a part of a course or project. I'm thinking of "calling" the script inside the actual selected template but I don't know if it's possible. For example, I create a new lecture note using the `select template` script, I select Lecture.md and inside Lecture.md there is a move funtction to move the note to Input Notes, then there is a call to my `move file to collection` script at the end of the template to check if it belongs to a collection or not and if so, choose the collection folder and move it there. If it does not belong to a collection, it will keep its current folder set by the previous move function within the selected template.

If you notice any problems, don't be shy!

`select template` template

```javascript
<%*
const templatesFolder = "Utilities/Templates";
const ttemplatesFolder = app.vault.getAbstractFileByPath(templatesFolder);

// Perform checks to ensure it's a folder and has subfolders
if (!("children" in ttemplatesFolder)) {
  throw new Error("Templates folder is not a folder.");
}

const templatesSubfolders = ttemplatesFolder.children.filter(subfolder => "children" in subfolder);

if (templatesSubfolders.length === 0) {
  throw new Error("Templates folder doesn't have any subfolders.");
}

let templatesCurrentFolder = ttemplatesFolder;
let selectedTemplatesFolder = "";

// Prompt user to select subfolder or go back
while (true) {
  const templatesSubfolders = templatesCurrentFolder.children.filter(subfolder => "children" in subfolder);
  const templatesFolderOptions = templatesSubfolders.map(subfolder => subfolder.name);

  // Check if it's the lowest subfolder
  if (templatesSubfolders.length === 0) {
    break;
  }

  const templatesOptions = ["Go Back", ...templatesFolderOptions];

  // Prompt user to select subfolder or go back
  const templatesSelectedOption = await tp.system.suggester(option => option, templatesOptions);

  if (templatesSelectedOption === "Go Back") {
    if (templatesCurrentFolder.parent) {
      templatesCurrentFolder = templatesCurrentFolder.parent;
      selectedTemplatesFolder = templatesCurrentFolder.path;
    }
  } else {
    const selectedSubfolder = templatesSubfolders.find(subfolder => subfolder.name === templatesSelectedOption);
    if (selectedSubfolder) {
      templatesCurrentFolder = selectedSubfolder;
      selectedTemplatesFolder = templatesCurrentFolder.path;
      break; // Break out of the loop once a subfolder is selected
    }
  }
}

// Get the selected template folder
const selectedTemplateFolder = app.vault.getAbstractFileByPath(selectedTemplatesFolder);

// Get the template files from the selected template folder
const templateFiles = selectedTemplateFolder.children.filter(file => !("children" in file));
const templateOptions = templateFiles.map(file => file.name);

// Prompt user to select template file
const selectedTemplateFile = await tp.system.suggester(option => option, templateOptions, true, "Select a template file:");

// Get the selected template file
const selectedTemplate = await tp.file.find_tfile(selectedTemplateFile);

// Prompt the user to enter a new title for the note
const newNoteTitle = await tp.system.prompt("Note title:");

// Create a new note with the selected template content and the entered title
await tp.file.create_new(selectedTemplate, newNoteTitle, true);
-%>
```

`move file to collection` template

```javascript
<%*
// Prompt user if note belongs to a collection
const belongsToCollection = await tp.system.suggester(
  ["YES", "NO"],
  ["YES", "NO"],
  false,
  "Does this note belong to a collection (i.e., course/project-specific material)?"
);

if (belongsToCollection === "YES") {
  // Get details about the "Collections" folder
  const collectionsFolder = "Collections";
  const tCollectionsFolder = app.vault.getAbstractFileByPath(collectionsFolder);

  // Perform checks to ensure it's a folder and has subfolders
  if (!("children" in tCollectionsFolder)) {
    throw new Error("Collections folder is not a folder.");
  }

  const collectionSubfolders = tCollectionsFolder.children.filter(subfolder => "children" in subfolder);

  if (collectionSubfolders.length === 0) {
    throw new Error("Collections folder doesn't have any collections.");
  }

  let currentFolder = tCollectionsFolder;
  let selectedFolder = "";

  while (true) {
    const subfolders = currentFolder.children.filter(subfolder => "children" in subfolder);
    const folderOptions = subfolders.map(subfolder => subfolder.name);
    
    // Check if it's the lowest subfolder
    if (subfolders.length === 0) {
      break;
    }

    const options = ["Go Back", folderOptions, "Current Folder"];

    // Prompt user to select subfolder or go back
    const selectedOption = await tp.system.suggester(option => option, options);

    if (selectedOption === "Go Back") {
      if (currentFolder.parent) {
        currentFolder = currentFolder.parent;
      }
      continue;
    }

    selectedFolder = subfolders.find(subfolder => subfolder.name === selectedOption).path;
    currentFolder = app.vault.getAbstractFileByPath(selectedFolder);
  }

  // Move the note to the selected subfolder or current folder
  const targetFolder = selectedFolder || currentFolder.path;
  await tp.file.move(`${targetFolder}/${tp.file.title}`);
}
-%>
```

**Comment by Zachatoo** [comment]
You can run another template from within your template using `tp.file.include()`

```js
<% tp.file.include("[[Template1]]") %>
```

https://silentvoid13.github.io/Templater/internal-functions/internal-modules/file-module.html#tpfileincludeinclude_link-string--tfile

**Reply by daamiian** [reply]
Yes, I saw that but every time I use it it fails to load the template. I imagine it's another problem causing it but I have since given up on this issue. I will close it now.

**Reply by daamiian** [reply]
actually before I do that, maybe I am thinking of include wrong. Does include actually append the template or does it just run the commands inside that template? Because really what I want to do is append the entire template to an existing note with yaml. The citations plugin loads my metadata for all of my input notes, so I create the notes there. That plugin uses mustache syntax, and it is inserted like this

```
---
type: 
alias: {{titleShort}}
authors: [{{authorString}}]
publisher: {{containerTitle}}
date: {{year}}
pages: {{page}}
doi: {{DOI}}
url: {{URL}}
---
<% tp.file.cursor() %>
<%*
const templateTitle = "% Select Input Note";
const templateFile = await tp.file.find_tfile(templateTitle);
tp.file.include(templateFile);
-%>
```

I used the TFile instead because the "[[]]" notation would give me errors

I'd like to append an entire template here (for example my lecture template which purposefully doesn't have any frontmatter set for this reason. Am I chasing a dream?

**Reply by daamiian** [reply]
I found the issue with `tp.file.include`. 

For whatever reason with my script, putting it outside as it's own line (<% tp.file.include(...) %>) instead of within the script fixed the issue. Now I can create a note using the citations plugin, have my metadata from zotero, and load the correct template for each note type. After grooming it a bit I will make a post to share it.

**Reply by Zachatoo** [reply, marked as answer]
You can have `tp.file.include` work in an execution command like this

```js
<%*
// Many lines
tR += tp.file.include("[[file name]]");
%>
```
